### PR TITLE
Tests: session management

### LINKED ISSUE #226: Unit test session management
We should be able to mock up dependencies and write unit tests for [`benefits.core.session`](https://github.com/cal-itp/benefits/blob/dev/benefits/core/session.py)

Django testing overview: https://docs.djangoproject.com/en/3.2/topics/testing/overview/

Here's a cool (and fairly straightforward) example of mocking a Django environment/`settings.py` ahead of running tests: https://github.com/fsbraun/django-automations/blob/master/runtests.py

_(Note the call out to the Django [`test` command](https://docs.djangoproject.com/en/3.2/topics/testing/overview/#running-tests)):_

```python
from django.core.management import call_command

# settings.configure()

# django.setup()

failures = call_command("test", test_path, interactive=False, failfast=False, verbosity=2)
```

### PR DESCRIPTION
**Pull request by thekaveman**
Closes #226 

*(Decided to talk about #334 more / leave it out of this PR)*

## Test coverage

```bash
$ coverage report --include benefits/core/session.py
Name                       Stmts   Miss Branch BrPart  Cover
------------------------------------------------------------
benefits/core/session.py     165      0     30      0   100%
```

## Coverage HTML report

This PR adds a helper script `tests/pytest/coverage.sh` that:

1. Runs the tests with `coverage`
2. Generates an HTML report in the app's `static/` directory

The report can be viewed by launching the app and navigating  to `http://localhost:$DJANGO_LOCAL_PORT/static/htmlcov/index.html`

### CODE REVIEW

**Review by afeld — APPROVED** [review]
My general suggestion here is to avoid explicit type hints whenever possible, especially with return types. Instead, lean on inferred types whenever possible. This will avoid the issue of type hints being incorrect, either from a mistake or from getting out of sync with code changes.

I'm going to Approve so that if you disagree you can still merge; also happy to merge this as "much better than nothing" and we can worry about that kind of thing as a refactor.

**Comment by afeld** [comment]
Sorry, hold up; got distracted and didn't look at the second half of the pull request 😆 One moment.

File: tests/pytest/conftest.py [hunk 1]
```diff
@@ -35,6 +36,9 @@ def initialize_session(request):
     """Helper initializes a Django request object's session."""
 
     # https://stackoverflow.com/a/55530933/358804
-    middleware = SessionMiddleware(lambda x: x)
-    middleware.process_request(request)
+    middleware = [SessionMiddleware(lambda x: x), LocaleMiddleware(lambda x: x)]
```
Comment (afeld) [inline tests/pytest/conftest.py:39]: We should change the function name+docstring to reflect that it does more now.

File: tests/pytest/core/test_session.py [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+def test_enrollment_token_default(session_request):
+    assert session._ENROLLMENT_TOKEN not in session_request.session
+    assert session.enrollment_token(session_request) is None
+    assert session._ENROLLMENT_TOKEN_EXP not in session_request.session
+    assert session.enrollment_token_expiry(session_request) is None
+
+
+@pytest.mark.django_db
+def test_enrollment_token_valid(session_request):
+    # valid token expiring in the far future
```
Comment (afeld) [inline tests/pytest/core/test_session.py:77]: Should we also have a test for an expired one?

File: docs/getting-started/testing.md [hunk 3]
```diff
@@ -75,5 +75,14 @@ See `tests/cypress/package.json` for more cypress scripts.
...
 ```
+
+### Coverage report
+
+Run the helper script `tests/pytest/coverage.sh` to:
+
+1. Runs the tests with `coverage`
+2. Generates an HTML report in the app's `static/` directory
+
+The report can be viewed by launching the app and navigating to `http://localhost:$DJANGO_LOCAL_PORT/static/htmlcov/index.html`
```
Comment (afeld) [inline docs/getting-started/testing.md:88]: I'd combine this with the previous section: we can just stick with a single way to run tests.

**Review by afeld — CHANGES_REQUESTED** [review]

**Comment by afeld** [comment]
Should have said up front: This is very thorough, thanks for doing all of it!

**Comment by thekaveman** [comment]
> My general suggestion here is to avoid explicit type hints whenever possible, especially with return types. Instead, lean on inferred types whenever possible. This will avoid the issue of type hints being incorrect, either from a mistake or from getting out of sync with code changes.

Fair point. We had talked about this a while back as a developer experience kind of thing, especially as we move between different areas of the codebase or onboard new folks, having the extra info built into the IDE experience felt like a net plus.

And while it was "related enough" to include in this PR, it's also "unrelated enough" to not include and save for another day, another conversation.

**Comment by afeld** [comment]
Yep yep! To be clear, I'm definitely a fan of types showing up through the editor… my point is just that this pull request was adding some return types in places where it could already be inferred by [Pyright](https://github.com/microsoft/pyright). (I did some testing removing the return type then hovering over the function name to see if anything changed.)

File: tests/pytest/core/test_session.py (hunk 2 above)
Comment (thekaveman) [reply tests/pytest/core/test_session.py]: Done in [bcf5562](https://github.com/cal-itp/benefits/pull/474/commits/bcf556238953f05a18126742c0dac9789e41991b)

File: docs/getting-started/testing.md (hunk 3 above)
Comment (thekaveman) [reply docs/getting-started/testing.md]: Done in [fdbd9ac](https://github.com/cal-itp/benefits/pull/474/commits/fdbd9ac4ea73b46ad713852cb58c79535a5f9af9)

File: tests/pytest/conftest.py (hunk 1 above)
Comment (thekaveman) [reply tests/pytest/conftest.py:41]: Done in [bb12659](https://github.com/cal-itp/benefits/pull/474/commits/bb12659b8be8436869c9c806851d75c29a459e35)

File: benefits/core/middleware.py [hunk 4]
```diff
@@ -38,7 +38,7 @@ def process_request(self, request):
             return None
 
         if request.method in settings.RATE_LIMIT_METHODS:
-            session.increment_rate_limit_counter(request)
+            session.rate_limit_counter_increment(request)
```
[the remark this answers is not in the document]
Comment (thekaveman) [reply benefits/core/middleware.py]: 👍 [325a004](https://github.com/cal-itp/benefits/pull/474/commits/325a00460c4e5bfc63aafba6076271f1f276f62f)

**Review by afeld — APPROVED** [review]

**Comment by afeld** [comment]
Merging in so I can build off of it.

**Pull request merged**